app/src/main/java/com/cornellappdev/transit/ui/viewmodels/HomeViewModel.kt

```diff
@@ -82,14 +110,53 @@ class HomeViewModel @Inject constructor(
      * Change the query in the search bar and update search results
      */
     fun onQueryChange(query: String) {
-        searchQuery.value = query;
+        if (query == "") {
+            _searchBarUiState.value = SearchBarUIState.RecentAndFavorites(
+                userPreferenceRepository.favoritesFlow.value,
+                userPreferenceRepository.recentsFlow.value
+            )
+        } else {
+            _searchBarUiState.value = SearchBarUIState.Query(
+                ApiResponse.Pending, query
+            )
+        }
+        routeRepository.makeSearch(query)
     }
 
     /**
      * Change the query in the add favorites search bar and update search results
      */
     fun onAddQueryChange(query: String) {
-        addSearchQuery.value = query;
+        addSearchQuery.value = query
+        routeRepository.makeSearch(query)
+    }
+
+    /**
+     * Asynchronous function to add a stop to recents
+     */
+    fun addRecent(stop: Place?) {
+        if (stop != null) {
+            var currentRecents = userPreferenceRepository.recentsFlow.value.toMutableList()
+            currentRecents = currentRecents
+                .filter { it != stop }
+                .toMutableList()
+            currentRecents.add(0, stop)
+            if (currentRecents.size > 5) {
+                currentRecents = currentRecents.take(5).toMutableList()
+            }
+            viewModelScope.launch {
+                userPreferenceRepository.setRecents(currentRecents)
+            }
+        }
+    }
```

Comment:
This seems like pretty Model-level logic. Say for example on some different screen you also want to add a recently searched stop. You'd have to copy this over to that VM, which creates code reuse.

For that reason, I think you should provide this as a function in `userPreferenceRepository` that you simply call here. That way, any other screens that may also want to add a recently searched stop (and thus go through all this nontrivial `take(5)` logic...) can simply call that function too.